<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین 4<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 style="color:rgb(90, 255, 184); font-size: 20px;">Comparing BART and GPT-2 for Text-to-SQL</h2>
<p align="center" style="color: #666; font-size: 16px;">زهرا فیض اله زاده</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">zahra.feizolahzadeh@gmail.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: کسری کاشانی</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: 810101490</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال: 1404/9/12</p>
</div>
</div>

<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="line-height: 2.0; font-size: 17px; color: black; font-family: Vazir;">
<div style="padding-right:40px">
</div>
<br>
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<br>
<li><b>سوال اول - <span dir="ltr">Text-to-SQL with BART and GPT-2</span> (50)</b></li>
<ul>
<li>بخش اول: بارگذاری و پیش‌پردازش داده‌های Text-to-SQL</li>
<li>بخش دوم: آموزش مدل Encoder–Decoder (BART)</li>
<li>بخش سوم: آموزش مدل Decoder-only (GPT-2)</li>
<li>بخش چهارم: مقایسه و تحلیل نهایی</li>

</ul>

</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
<li>	مسئله Text-to-SQL حساسیت بسیار بالایی به جزئیات دارد.</li>
کوچک‌ترین اختلاف در ساختار SQL، ترتیب کلیدواژه‌ها یا حتی فاصله‌ها می‌تواند باعث نامعتبر شدن خروجی شود. بنابراین هنگام ارزیابی، به‌ویژه در معیار Normalized EM، باید دقت زیادی به اصلاح فاصله‌ها، حذف سمیکالن اضافه، و یکسان‌سازی case کلمات داشته باشید تا مقایسه‌ای منصفانه انجام شود.
<li>	معماری‌های Encoder–Decoder و Decoder-only رفتارهای متفاوتی در تولید SQL دارند.</li>
ساخت صحیح ورودی‌ها (Prefix)، نحوه قرار دادن question و schema در متن، ترتیب tokenها،          padding مناسب و mask کردن بخش‌های غیرهدف، همگی نقش اساسی در عملکرد این مدل‌ها دارند. دقت کنید که GPT-2 باید تمام اطلاعات را در قالب یک توالی یکپارچه دریافت کند، درحالی‌که BART ورودی را رمزگذاری و سپس خروجی SQL را مستقل تولید می‌کند.
<li>	برای مقایسه مدل‌ها، آموزش حداقلی اما واقعی لازم است.</li>
باید حداقل یک ایپاک کامل روی حداقل ۲۰۰۰۰ نمونه از داده‌های آموزش اجرا شود تا مدل به‌طور معنی‌دار الگوهای اولیه Text-to-SQL را یاد بگیرد و ارزیابی‌ها قابل اعتماد باشند. لازم نیست آموزش طولانی انجام دهید، اما اجرای همین مقدار حداقلی برای تحلیل و مقایسه معماری‌ها ضروری است.
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>سوال اول: تحلیل و مقایسه معماری‌های Encoder–Decoder و Decoder-Only در تسک Text-to-SQL <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش از تمرین با مسئله Text-to-SQL آشنا می‌شوید؛ وظیفه‌ای که در آن مدل باید پرسش‌های زبان طبیعی را به کوئری SQL معتبر تبدیل کند. هدف تمرین، مقایسه عملی دو معماری مهم ترنسفورمری است: Encoder–Decoder‌ (مانند BART) و Decoder-only‌ (مانند GPT-2). پس از آماده‌سازی داده‌ها، ابتدا مدل BART و سپس مدل GPT-2 را برای این تسک آموزش می‌دهید و در نهایت عملکرد دو مدل را از نظر دقت، رفتار تولیدی و انواع خطاها تحلیل و مقایسه خواهید کرد.
</p>




## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش اول: مجموعه‌داده و پیش‌پردازش(10 نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
داده‌های Gretel Synthetic Text-to-SQL را از HuggingFace بارگذاری کرده و فایل‌های train/test را دریافت کنید. ساختار دیتاست را بررسی کرده، یک نمونه نمایش دهید و نقش ستون‌های اصلی از جمله question، schema و query را توضیح دهید. سپس داده‌ها را با استفاده از train_test_split به بخش‌های train و dev تقسیم کنید و ستون‌ها را به اسامی استاندارد (question، schema، query) تغییر دهید. برای کاهش زمان آموزش، روی هر بخش محدودیت اندازه تعیین کنید. در پایان، تحلیل کوتاهی از ماهیت پرسش‌ها، ساختار schemaها و چالش‌هایی که برای مدل ایجاد می‌کنند ارائه دهید.
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
در پایان این بخش انتظار می‌رود دانشجو یک دیتافریم کاملا آماده برای مراحل بعدی در اختیار داشته باشد. خروجی مطلوب شامل بارگذاری صحیح فایل‌های train و test، نمایش ساختار کلی داده‌ها، تقسیم درست داده‌ها به بخش‌های train و dev، تغییر نام ستون‌ها به قالب استاندارد، و اعمال زیرنمونه‌گیری برای محدود کردن حجم داده‌ها است. علاوه بر این، باید یک نمونه واقعی از داده‌ها همراه با توضیح مختصر درباره ماهیت سؤال، ساختار schema و نمونه SQL نمایش داده شود تا مشخص شود داده‌ها به‌درستی پردازش شده و برای ورود به مراحل آموزش مدل آماده هستند.

</p>

In [1]:
import pandas as pd
from datasets import load_dataset
from sklearn.model_selection import train_test_split

In [2]:
dataset = load_dataset("gretelai/synthetic_text_to_sql")

print(dataset)

README.md: 0.00B [00:00, ?B/s]

synthetic_text_to_sql_train.snappy.parqu(…):   0%|          | 0.00/32.4M [00:00<?, ?B/s]

synthetic_text_to_sql_test.snappy.parque(…):   0%|          | 0.00/1.90M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/100000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/5851 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
        num_rows: 100000
    })
    test: Dataset({
        features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
        num_rows: 5851
    })
})


In [3]:
train_dataset = dataset["train"]
test_dataset = dataset["test"]

print(train_dataset)
print(test_dataset)

Dataset({
    features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
    num_rows: 100000
})
Dataset({
    features: ['id', 'domain', 'domain_description', 'sql_complexity', 'sql_complexity_description', 'sql_task_type', 'sql_task_type_description', 'sql_prompt', 'sql_context', 'sql', 'sql_explanation'],
    num_rows: 5851
})


In [4]:
df_train_full = train_dataset.to_pandas()
df_test_full = test_dataset.to_pandas()

In [5]:
df_train_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype 
---  ------                      --------------   ----- 
 0   id                          100000 non-null  int32 
 1   domain                      100000 non-null  object
 2   domain_description          100000 non-null  object
 3   sql_complexity              100000 non-null  object
 4   sql_complexity_description  100000 non-null  object
 5   sql_task_type               100000 non-null  object
 6   sql_task_type_description   100000 non-null  object
 7   sql_prompt                  100000 non-null  object
 8   sql_context                 100000 non-null  object
 9   sql                         100000 non-null  object
 10  sql_explanation             100000 non-null  object
dtypes: int32(1), object(10)
memory usage: 8.0+ MB


In [6]:
df_train_full

id                 domain  \
0       5097               forestry   
1       5098       defense industry   
2       5099         marine biology   
3       5100     financial services   
4       5101                 energy   
...      ...                    ...   
99995  89651              nonprofit   
99996  89652                 retail   
99997  89653       fitness industry   
99998  89654      space exploration   
99999  89655  wildlife conservation   

                                      domain_description    sql_complexity  \
0      Comprehensive data on sustainable forest manag...       single join   
1      Defense contract data, military equipment main...       aggregation   
2      Comprehensive data on marine species, oceanogr...         basic SQL   
3      Detailed financial data including investment s...       aggregation   
4      Energy market data covering renewable energy s...  window functions   
...                                                  ...               ...   
99995  Nonprofit data on charitable giving trends, so...         basic SQL   
99996  Retail data on circular supply chains, ethical...       single join   
99997  Workout data, membership demographics, wearabl...       single join   
99998  Spacecraft manufacturing data, space mission r...       single join   
99999  Animal population data, habitat preservation e...         basic SQL   

                              sql_complexity_description  \
0            only one join (specify inner, outer, cross)   
1      aggregation functions (COUNT, SUM, AVG, MIN, M...   
2               basic SQL with a simple select statement   
3      aggregation functions (COUNT, SUM, AVG, MIN, M...   
4      window functions (e.g., ROW_NUMBER, LEAD, LAG,...   
...                                                  ...   
99995           basic SQL with a simple select statement   
99996        only one join (specify inner, outer, cross)   
99997        only one join (specify inner, outer, cross)   
99998        only one join (specify inner, outer, cross)   
99999           basic SQL with a simple select statement   

                 sql_task_type  \
0      analytics and reporting   
1      analytics and reporting   
2      analytics and reporting   
3      analytics and reporting   
4      analytics and reporting   
...                        ...   
99995  analytics and reporting   
99996  analytics and reporting   
99997  analytics and reporting   
99998  analytics and reporting   
99999  analytics and reporting   

                               sql_task_type_description  \
0      generating reports, dashboards, and analytical...   
1      generating reports, dashboards, and analytical...   
2      generating reports, dashboards, and analytical...   
3      generating reports, dashboards, and analytical...   
4      generating reports, dashboards, and analytical...   
...                                                  ...   
99995  generating reports, dashboards, and analytical...   
99996  generating reports, dashboards, and analytical...   
99997  generating reports, dashboards, and analytical...   
99998  generating reports, dashboards, and analytical...   
99999  generating reports, dashboards, and analytical...   

                                              sql_prompt  \
0      What is the total volume of timber sold by eac...   
1      List all the unique equipment types and their ...   
2      How many marine species are found in the South...   
3      What is the total trade value and average pric...   
4      Find the energy efficiency upgrades with the h...   
...                                                  ...   
99995  Which programs had the highest volunteer parti...   
99996  What is the number of fair-trade certified acc...   
99997  Find the user with the longest workout session...   
99998  How many space missions were completed by each...   
99999  Determine the number of unique animal species ...   

                                

In [7]:
df_test_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5851 entries, 0 to 5850
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          5851 non-null   int32 
 1   domain                      5851 non-null   object
 2   domain_description          5851 non-null   object
 3   sql_complexity              5851 non-null   object
 4   sql_complexity_description  5851 non-null   object
 5   sql_task_type               5851 non-null   object
 6   sql_task_type_description   5851 non-null   object
 7   sql_prompt                  5851 non-null   object
 8   sql_context                 5851 non-null   object
 9   sql                         5851 non-null   object
 10  sql_explanation             5851 non-null   object
dtypes: int32(1), object(10)
memory usage: 480.1+ KB


In [8]:
df_test_full

id                   domain  \
0        1  artificial intelligence   
1        2        rural development   
2        3                aerospace   
3        4              aquaculture   
4        5   entertainment industry   
...    ...                      ...   
5846  5847                  museums   
5847  5848         waste management   
5848  5849          water resources   
5849  5850         fitness industry   
5850  5851         fitness industry   

                                     domain_description sql_complexity  \
0     AI data on algorithmic fairness, AI safety, ex...      basic SQL   
1     Agricultural innovation metrics, rural infrast...      basic SQL   
2     Aircraft manufacturing data, satellite deploym...    aggregation   
3     Aquatic farming data, fish stock management, o...      basic SQL   
4     Audience demographics, viewership trends, cont...      basic SQL   
...                                                 ...            ...   
5846  Visitor demographics, exhibition analytics, co...      basic SQL   
5847  Waste generation metrics, recycling rates, lan...      basic SQL   
5848  Water usage metrics, drought impact assessment...    aggregation   
5849  Workout data, membership demographics, wearabl...    aggregation   
5850  Workout data, membership demographics, wearabl...     subqueries   

                             sql_complexity_description  \
0              basic SQL with a simple select statement   
1              basic SQL with a simple select statement   
2     aggregation functions (COUNT, SUM, AVG, MIN, M...   
3              basic SQL with a simple select statement   
4              basic SQL with a simple select statement   
...                                                 ...   
5846           basic SQL with a simple select statement   
5847           basic SQL with a simple select statement   
5848  aggregation functions (COUNT, SUM, AVG, MIN, M...   
5849  aggregation functions (COUNT, SUM, AVG, MIN, M...   
5850  subqueries, including correlated and nested su...   

                sql_task_type  \
0     analytics and reporting   
1           data manipulation   
2     analytics and reporting   
3     analytics and reporting   
4     analytics and reporting   
...                       ...   
5846  analytics and reporting   
5847  analytics and reporting   
5848  analytics and reporting   
5849  analytics and reporting   
5850  analytics and reporting   

                              sql_task_type_description  \
0     generating reports, dashboards, and analytical...   
1              inserting, updating, or deleting records   
2     generating reports, dashboards, and analytical...   
3     generating reports, dashboards, and analytical...   
4     generating reports, dashboards, and analytical...   
...                                                 ...   
5846  generating reports, dashboards, and analytical...   
5847  generating reports, dashboards, and analytical...   
5848  generating reports, dashboards, and analytical...   
5849  generating reports, dashboards, and analytical...   
5850  generating reports, dashboards, and analytical...   

                                             sql_prompt  \
0     What is the average explainability score of cr...   
1     Delete all records of rural infrastructure pro...   
2     How many accidents have been recorded for Spac...   
3     What is the maximum quantity of seafood sold i...   
4     What is the total budget for movies released b...   
...                                                 ...   
5846  How many visitors are from the city of "Seattl...   
5847  What is the total waste generation in kilogram...   
5848  What is the maximum wastewater volume treated ...   
5849  Calculate the total workout duration and numbe...   
5850  How many users have a higher heart rate on wee...   

                                            sql_context  \
0     CREATE TABLE creative_ai (application_id INT, ...   
1     CR

In [9]:
col_map = {
    "sql_prompt": "question",
    "sql_context": "schema",
    "sql": "query"
}

df_train_full = df_train_full.rename(columns=col_map)
df_test_full = df_test_full.rename(columns=col_map)

In [10]:
df_train_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 11 columns):
 #   Column                      Non-Null Count   Dtype 
---  ------                      --------------   ----- 
 0   id                          100000 non-null  int32 
 1   domain                      100000 non-null  object
 2   domain_description          100000 non-null  object
 3   sql_complexity              100000 non-null  object
 4   sql_complexity_description  100000 non-null  object
 5   sql_task_type               100000 non-null  object
 6   sql_task_type_description   100000 non-null  object
 7   question                    100000 non-null  object
 8   schema                      100000 non-null  object
 9   query                       100000 non-null  object
 10  sql_explanation             100000 non-null  object
dtypes: int32(1), object(10)
memory usage: 8.0+ MB


In [11]:
df_test_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5851 entries, 0 to 5850
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          5851 non-null   int32 
 1   domain                      5851 non-null   object
 2   domain_description          5851 non-null   object
 3   sql_complexity              5851 non-null   object
 4   sql_complexity_description  5851 non-null   object
 5   sql_task_type               5851 non-null   object
 6   sql_task_type_description   5851 non-null   object
 7   question                    5851 non-null   object
 8   schema                      5851 non-null   object
 9   query                       5851 non-null   object
 10  sql_explanation             5851 non-null   object
dtypes: int32(1), object(10)
memory usage: 480.1+ KB


In [12]:
df_train, df_dev = train_test_split(
    df_train_full,
    test_size=0.1,
    random_state=42,
    shuffle=True,
)

In [13]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 90000 entries, 51994 to 15795
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          90000 non-null  int32 
 1   domain                      90000 non-null  object
 2   domain_description          90000 non-null  object
 3   sql_complexity              90000 non-null  object
 4   sql_complexity_description  90000 non-null  object
 5   sql_task_type               90000 non-null  object
 6   sql_task_type_description   90000 non-null  object
 7   question                    90000 non-null  object
 8   schema                      90000 non-null  object
 9   query                       90000 non-null  object
 10  sql_explanation             90000 non-null  object
dtypes: int32(1), object(10)
memory usage: 7.9+ MB


In [14]:
df_dev.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10000 entries, 75721 to 74479
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          10000 non-null  int32 
 1   domain                      10000 non-null  object
 2   domain_description          10000 non-null  object
 3   sql_complexity              10000 non-null  object
 4   sql_complexity_description  10000 non-null  object
 5   sql_task_type               10000 non-null  object
 6   sql_task_type_description   10000 non-null  object
 7   question                    10000 non-null  object
 8   schema                      10000 non-null  object
 9   query                       10000 non-null  object
 10  sql_explanation             10000 non-null  object
dtypes: int32(1), object(10)
memory usage: 898.4+ KB


In [15]:
SIZE_TRAIN = 20000
SIZE_DEV = 500
SIZE_TEST = 2000

df_train = df_train.sample(n=SIZE_TRAIN, random_state=42).reset_index(drop=True)
df_dev = df_dev.sample(n=SIZE_DEV, random_state=42).reset_index(drop=True)
df_test = df_test_full.sample(n=SIZE_TEST, random_state=42).reset_index(drop=True)

In [16]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          20000 non-null  int32 
 1   domain                      20000 non-null  object
 2   domain_description          20000 non-null  object
 3   sql_complexity              20000 non-null  object
 4   sql_complexity_description  20000 non-null  object
 5   sql_task_type               20000 non-null  object
 6   sql_task_type_description   20000 non-null  object
 7   question                    20000 non-null  object
 8   schema                      20000 non-null  object
 9   query                       20000 non-null  object
 10  sql_explanation             20000 non-null  object
dtypes: int32(1), object(10)
memory usage: 1.6+ MB


In [17]:
df_dev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          500 non-null    int32 
 1   domain                      500 non-null    object
 2   domain_description          500 non-null    object
 3   sql_complexity              500 non-null    object
 4   sql_complexity_description  500 non-null    object
 5   sql_task_type               500 non-null    object
 6   sql_task_type_description   500 non-null    object
 7   question                    500 non-null    object
 8   schema                      500 non-null    object
 9   query                       500 non-null    object
 10  sql_explanation             500 non-null    object
dtypes: int32(1), object(10)
memory usage: 41.1+ KB


In [18]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   id                          2000 non-null   int32 
 1   domain                      2000 non-null   object
 2   domain_description          2000 non-null   object
 3   sql_complexity              2000 non-null   object
 4   sql_complexity_description  2000 non-null   object
 5   sql_task_type               2000 non-null   object
 6   sql_task_type_description   2000 non-null   object
 7   question                    2000 non-null   object
 8   schema                      2000 non-null   object
 9   query                       2000 non-null   object
 10  sql_explanation             2000 non-null   object
dtypes: int32(1), object(10)
memory usage: 164.2+ KB


In [19]:
example_row = df_train.sample(1, random_state=42).iloc[0]

print("--- An example row from training dataset ---")
print("\nQuestion:\n", example_row["question"])
print("\nSchema:\n", example_row["schema"])
print("\nQuery:\n", example_row["query"])

--- An example row from training dataset ---

Question:
 What is the maximum fare for buses in Tokyo?

Schema:
 CREATE TABLE if not exists bus_fares (id INT, city VARCHAR(20), max_fare DECIMAL(3,2)); INSERT INTO bus_fares (id, city, max_fare) VALUES (1, 'Tokyo', 3.50), (2, 'Seoul', 2.80);

Query:
 SELECT max_fare FROM bus_fares WHERE city = 'Tokyo';


<div dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی:</b><br>
<br>
ماهیت question:
<br>
طبق سوالی که در نمونه مشاهده می‌شود، این یک سوال factual ساده است که به دنبال یک مقدار عددی مشخص می‌گردد. از منظر SQL، بیشتر شبیه یک SELECT با شرط WHERE است تا یک aggregation پیچیده. با این حال، عبارت maximum fare از نظر معنایی یادآور MAX() هم هست، هرچند در این دیتابیس، نام خود ستون max_fare است.
<br>
عبارت maximum fare یعنی باید با ستون max_fare مپ شود نه مثلا id یا city.
<br>
عبارت for buses یعنی اشاره به جدول bus_fares دارد که خود واژه‌ی bus در نام جدول وجود دارد.
<br>
عبارت in Tokyo یعنی باید تشخیص دهد که Tokyo یک مقدار ('Tokyo') برای ستون city است، نه نام جدول یا دیتابیس.
<br>
چالش‌هایی که طبق همین یک نمونه برای مدل ممکن است بوجود بیاید، این است که مدل باید بفهمد maximum fare در این context یعنی ستون max_fare. همچنین باید تشخیص دهد که پاسخ، یک ستون از جدول است، نه count یا max. همچنین باید بفهمذ که شهر Tokyo یک filter روی ستون city ایجاد می‌کند.
<br>
برای ارزیابی pipeline داده، این سوال نشان می‌دهد که ستون question متن طبیعی و قابل تفسیر برای مدل را در فرمت تمیز در خودش نگه داشته است؛ و تگ اضافه‌ یا فرمت خرابی ندارد.
<br>

<br>
ساختار schema:
<br>
schema به صورت یک string از SQL است، نه یک ساختار JSON:

CREATE TABLE if not exists bus_fares (

id INT,

city VARCHAR(20),

max_fare DECIMAL(3,2)

);

INSERT INTO bus_fares (id, city, max_fare) VALUES

(1, 'Tokyo', 3.50),

(2, 'Seoul', 2.80);


<br>
در بخش CREATE TABLE، اطلاعاتی که به مدل داده می‌شود این است که نام جدول bus_fares است، ستون‌هایش id از نوع INT، city از نوع VARCHAR(20) که رشته‌ای و محل مناسبی برای ذخیره‌ی نام شهر است، max_fare از نوع DECIMAL(3,2) که مقدار عددی اعشاری و هم‌خوان با fare در واقعیت است.
<br>
در بخش INSERT، دو ردیف نمونه‌ی (Tokyo, 3.50) و (Tokyo, 3.50) به جدول اضافه می‌شوند که برای Text-to-SQL، این قسمت از نظر semantics معمولا نقش مثال را دارد و کمک می‌کند مدل بفهمد که city مقادیرش اسم شهر است و max_fare نیز عدد است.
<br>
چالش‌های schema برای مدل نیز این است که schema به صورت رشته‌ی SQL خام است، و نه یک لیست تمیز از جدول‌ها و ستون‌ها. لذا مدل باید خودش بین نام ستون‌ها و مقادیر فرق بگذارد، بین ساختار پایگاه داده (bus_fares، city، max_fare) و داده‌های نمونه تمایز ایجاد کند، و از وسط این شلوغی، همان چند تکه‌ی مهم را برای تولید query استخراج کند.
<br>
از نگاه pipeline داده نیز، وجود چنین schemaای در ستون schema نشان می‌دهد که دیتاست درست لود شده و schema به شکل کامل و خام، نه ناقص یا truncated، در اختیار مدل قرار می‌گیرد.
<br>

<br>
تحلیل query:
<br>
این query داده‌شده دقیقا پاسخ سوال را از روی schema استخراج می‌کند:

SELECT max_fare

FROM bus_fares

WHERE city = 'Tokyo';

<br>
برای maximum fare در سوال، ستون max_fare را در SELECT قرار داده است.
<br>
برای for buses در سوال، به جدول bus_fares در FROM اشاره کرده است.
<br>
برای in Tokyo در سوال، شرط city = 'Tokyo' را در WHERE گذاشته است.
<br>
لذا در اینجا مدل باید سه نگاشت اساسی را درست انجام دهد: mapping عبارت natural language به ستون صحیح، mapping اشاره به entity متنی به مقدار فیلتر، و تشخیص جدول مرتبط.
<br>
این query از نظر syntax ساده است، ولی مثلا اگر مدل کوتیشن‌ها را جا بیندازد، query از نظر SQL نامعتبر می‌شود. یا مثلا اگر WHERE را جا بیندازد، نتیجه از نظر منطق ساال اشتباه می‌شود. همچنین اگر از MAX() استفاده کند، از نظر معنایی درست است اما از نظر مقایسه‌ی EM ممکن است به عنوان اشتباه شمرده شود، مگر اینکه در معیار نرمال‌سازی چنین تفاوت‌هایی نیز لحاظ شود.
<br>

<br>
در نتیجه، نمایش همین یک نمونه نشان می‌دهد که pipeline بارگذاری و پیش‌پردازش داده‌ها به‌درستی عمل کرده است و متن سوال، متن schema و query SQL مربوطه در سه ستون استاندارد question، schema و query ذخیره شده‌اند و هم‌خوانی معنایی کامل دارند. بنابراین داده‌ها در قالبی قرار گرفته‌اند که مستقیما برای آموزش مدل‌های Text-to-SQL (چه Encoder–Decoder و چه Decoder-only) قابل استفاده است.
<br>

<br>
همچنین، همین مثال ساده چند چالش کلاسیک Text-to-SQL را نشان می‌دهد. مثلا درک زبانی برای ساختار SQL، استخراج اطلاعات از schema متنی، تولید SQL نحوی‌ معتبر، و نیز تفاوت مدل‌ها
</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش دوم: آموزش مدل Encoder–Decoder (BART) (۱۵ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش از تمرین، وارد مرحله‌ی پیاده‌سازی مدل BART می‌شویم. پیش از شروع، لازم است به‌صورت مختصر و دقیق توضیح دهید که مدل BART چیست و چگونه کار می‌کند؛ به‌ویژه این‌که چرا معماری Encoder–Decoder آن برای وظایفی مانند Text-to-SQL مناسب است. انتظار می‌رود دانشجو به ساختار دو‌بخشی BART، نحوهٔ پردازش ورودی در انکودر، و تولید خروجی در دیکودر اشاره کند و نشان دهد که این معماری برای نگاشت دنباله‌ای از کلمات (سؤال + schema) به دنباله‌ای دیگر (SQL) کاملاً مناسب است.
پس از توضیح مدل، باید قالب ورودی مناسبی طراحی کنید که در آن سؤال و schema به‌صورت یک رشتهٔ یکپارچه در اختیار مدل قرار گیرند. سپس یک Dataset اختصاصی برای BART بسازید که وظایفی مانند tokenize کردن ورودی، ساخت attention mask‌ها و آماده‌سازی SQL به عنوان خروجی هدف مدل را انجام دهد. مدل را حداقل یک ایپاک روی ۲۰۰۰۰ نمونه آموزش دهید و loss نهایی، زمان اجرا و هایپرپارامترهای اصلی را گزارش کنید.
در مرحلهٔ ارزیابی، باید عملکرد مدل را روی dev و test با دو معیار Raw Exact Match و Normalized Exact Match بسنجید.
از شما انتظار می‌رود که هر دو معیار را توضیح دهید:
Raw EM چگونه مقایسهٔ مستقیم دو رشته را انجام می‌دهد، چرا ممکن است برای SQL ناکافی باشد، و Normalized EM چگونه با حذف تفاوت‌های سطحی (مثل فاصله‌ها، علامت‌ها و بزرگی یا کوچکی حروف) مقایسه‌ای عادلانه‌تر ارائه می‌دهد.
در پایان، پنج نمونهٔ تصادفی از dev انتخاب کنید و پرسش، خروجی مدل، SQL طلایی و نوع خطاها را تحلیل کنید.
<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
نکته مهم:
برای پیاده‌سازی Raw EM می‌توانید از مقایسهٔ مستقیم رشته‌ها در پایتون استفاده کنید.
برای Normalized EM مجاز هستید از ابزارهایی مانند sqlparse برای فرمت‌بندی SQL و توابع سادهٔ پردازش متن جهت lowercase کردن و حذف فاصله‌های اضافی استفاده کنید، اما منطق مقایسه‌ی Normalized باید توسط خودتان نوشته شود.
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک مدل BART آموزش‌دیده و ارزیابی‌شده است که بتواند از روی پرسش و ساختار پایگاه‌داده، کوئری SQL معتبر تولید کرده و معیارهای Raw EM و Normalized EM آن روی داده‌های dev و test گزارش شده باشد.
</p>

In [20]:
import re
import time
import torch
import random
import sqlparse
import numpy as np
from torch.utils.data import Dataset
from transformers import Seq2SeqTrainingArguments, TrainingArguments, Seq2SeqTrainer, Trainer, DataCollatorForSeq2Seq, default_data_collator, BartTokenizer, BartForConditionalGeneration

2025-12-04 00:26:38.747393: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1764807998.917579      47 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1764807998.970781      47 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


AttributeError: 'MessageFactory' object has no attribute 'GetPrototype'

AttributeError: 'MessageFactory' object has no attribute 'GetPrototype'

AttributeError: 'MessageFactory' object has no attribute 'GetPrototype'

AttributeError: 'MessageFactory' object has no attribute 'GetPrototype'

AttributeError: 'MessageFactory' object has no attribute 'GetPrototype'

In [21]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

print(device)

cuda


In [22]:
class TextToSQLBartDataset(Dataset):
    def __init__(self, df, tokenizer, max_source_len=512, max_target_len=256):
        self.df = df.reset_index(drop=True)

        self.tokenizer = tokenizer

        self.max_source_len = max_source_len

        self.max_target_len = max_target_len

    def __len__(self):
        return len(self.df)

    def _build_prompt(self, question, schema):
        return f"Question: {question}\nSchema: {schema}"

    def __getitem__(self, idx):
        row = self.df.iloc[idx]

        question = row["question"]
        schema = row["schema"]
        query = row["query"]

        source_text = self._build_prompt(question, schema)
        target_text = query

        source_for_encoder = self.tokenizer(
            source_text,
            max_length=self.max_source_len,
            truncation=True,
            padding="max_length",
            return_tensors="pt",
        )

        target_for_encoder = self.tokenizer(
            target_text,
            max_length=self.max_target_len,
            truncation=True,
            padding="max_length",
            return_tensors="pt",
        )

        input_ids = source_for_encoder["input_ids"].squeeze(0)
        attention_mask = source_for_encoder["attention_mask"].squeeze(0)
        labels = target_for_encoder["input_ids"].squeeze(0)

        labels[labels == self.tokenizer.pad_token_id] = -100

        return {
            "input_ids": input_ids,
            "attention_mask": attention_mask,
            "labels": labels
        }

In [23]:
MODEL_NAME_BART = "facebook/bart-base"

tokenizer_bart = BartTokenizer.from_pretrained(MODEL_NAME_BART)

model_bart = BartForConditionalGeneration.from_pretrained(MODEL_NAME_BART)
model_bart.to(device)

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/558M [00:00<?, ?B/s]

BartForConditionalGeneration(
  (model): BartModel(
    (shared): BartScaledWordEmbedding(50265, 768, padding_idx=1)
    (encoder): BartEncoder(
      (embed_tokens): BartScaledWordEmbedding(50265, 768, padding_idx=1)
      (embed_positions): BartLearnedPositionalEmbedding(1026, 768)
      (layers): ModuleList(
        (0-5): 6 x BartEncoderLayer(
          (self_attn): BartAttention(
            (k_proj): Linear(in_features=768, out_features=768, bias=True)
            (v_proj): Linear(in_features=768, out_features=768, bias=True)
            (q_proj): Linear(in_features=768, out_features=768, bias=True)
            (out_proj): Linear(in_features=768, out_features=768, bias=True)
          )
          (self_attn_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
          (activation_fn): GELUActivation()
          (fc1): Linear(in_features=768, out_features=3072, bias=True)
          (fc2): Linear(in_features=3072, out_features=768, bias=True)
          (final_layer_n

In [24]:
MAX_SOURCE_LEN = 512
MAX_TARGET_LEN = 256

dataset_train = TextToSQLBartDataset(df_train, tokenizer_bart, MAX_SOURCE_LEN, MAX_TARGET_LEN)
dataset_dev = TextToSQLBartDataset(df_dev, tokenizer_bart, MAX_SOURCE_LEN, MAX_TARGET_LEN)
dataset_test = TextToSQLBartDataset(df_test, tokenizer_bart, MAX_SOURCE_LEN, MAX_TARGET_LEN)

In [62]:
BATCH_SIZE = 4
TRAIN_EPOCHS = 1
LEARNING_RATE = 5e-5

training_args = Seq2SeqTrainingArguments(
    output_dir="./bart_text2sql",
    num_train_epochs=TRAIN_EPOCHS,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=BATCH_SIZE,
    learning_rate=LEARNING_RATE,
    weight_decay=0.01,
    warmup_ratio=0.05,
    fp16=False,
    save_total_limit=1,
    logging_steps=100,
    report_to="none"
)

In [26]:
data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer_bart, model=model_bart)

trainer = Seq2SeqTrainer(
    model=model_bart,
    args=training_args,
    train_dataset=dataset_train,
    eval_dataset=dataset_dev,
    data_collator=data_collator,
    tokenizer=tokenizer_bart
)

/tmp/ipykernel_47/581656082.py:3: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Seq2SeqTrainer.__init__`. Use `processing_class` instead.
  trainer = Seq2SeqTrainer(


In [27]:
start_time = time.time()
train_result = trainer.train()
end_time = time.time()

final_train_loss = train_result.training_loss
train_metrics = train_result.metrics

print("\n=== Training Report ===")
print(f"\nFinal training loss: {final_train_loss}")
print("\nTraining time (seconds):", end_time - start_time)
print("\nTrain metrics:", train_metrics)
print("\nHyperparameters:")
print(f"  num_train_epochs: {training_args.num_train_epochs}")
print(f"  per_device_train_batch_size: {training_args.per_device_train_batch_size}")
print(f"  max_source_len: {MAX_SOURCE_LEN}")
print(f"  max_target_len: {MAX_TARGET_LEN}")
print(f"  learning_rate: {training_args.learning_rate}")
print(f"  weight_decay: {training_args.weight_decay}")
print(f"  warmup_ratio: {training_args.warmup_ratio}")

/usr/local/lib/python3.11/dist-packages/transformers/data/data_collator.py:741: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at /pytorch/torch/csrc/utils/tensor_new.cpp:254.)
  batch["labels"] = torch.tensor(batch["labels"], dtype=torch.int64)


/usr/local/lib/python3.11/dist-packages/transformers/modeling_utils.py:3685: UserWarning: Moving the following attributes in the config to the generation config: {'early_stopping': True, 'num_beams': 4, 'no_repeat_ngram_size': 3, 'forced_bos_token_id': 0}. You are seeing this warning because you've set generation parameters in the model config, as opposed to in the generation config.
  warnings.warn(



=== Training Report ===

Final training loss: 0.7199405433654785

Training time (seconds): 2017.7950155735016

Train metrics: {'train_runtime': 2017.3907, 'train_samples_per_second': 9.914, 'train_steps_per_second': 2.478, 'total_flos': 6097364582400000.0, 'train_loss': 0.7199405433654785, 'epoch': 1.0}

Hyperparameters:
  num_train_epochs: 1
  per_device_train_batch_size: 4
  max_source_len: 512
  max_target_len: 256
  learning_rate: 5e-05
  weight_decay: 0.01
  warmup_ratio: 0.05


In [28]:
def normalize_sql(sql):
    if sql is None:
        return ""

    sql = sql.strip()

    formatted = sqlparse.format(
        sql,
        keyword_case="lower",
        identifier_case="lower",
        strip_comments=True,
        reindent=False
    )

    formatted = re.sub(r"\s+", " ", formatted)
    formatted = formatted.strip()

    if formatted.endswith(";"):
        formatted = formatted[ : -1].strip()

    return formatted

def compute_em_metrics(pred_sql, label_sql):
    if pred_sql is None:
        return 0, 0

    raw_em = int(pred_sql.strip() == label_sql.strip())
    norm_em = int(normalize_sql(pred_sql) == normalize_sql(label_sql))

    return raw_em, norm_em

In [29]:
def pred_sql(model, tokenizer, questions, schemas, max_source_len=512, max_target_len=256):
    source_texts = []
    for question, schema in zip(questions, schemas):
        source_texts.append(f"Question: {question}\nSchema: {schema}")

    sources_for_encoder = tokenizer(
        source_texts,
        max_length=max_source_len,
        padding=True,
        truncation=True,
        return_tensors="pt"
    ).to(model.device)

    input_ids = sources_for_encoder["input_ids"]
    attention_mask = sources_for_encoder["attention_mask"]

    with torch.no_grad():
        generated_ids = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_length=max_target_len,
            num_beams=4
        )

    pred_sqls = tokenizer.batch_decode(
        generated_ids,
        skip_special_tokens=True
    )

    return pred_sqls

In [30]:
def evaluate_bart(model, tokenizer, df, batch_size=8, max_source_len=512, max_target_len=256):
    raw_ems = []
    norm_ems = []

    for start in range(0, len(df), batch_size):
        end = min(start + batch_size, len(df))

        batch = df.iloc[start : end]

        questions = batch["question"].tolist()
        schemas = batch["schema"].tolist()
        queries = batch["query"].tolist()

        pred_sqls = pred_sql(model, tokenizer, questions, schemas, max_source_len, max_target_len)

        for pred, label in zip(pred_sqls, queries):
            raw_em, norm_em = compute_em_metrics(pred, label)

            raw_ems.append(raw_em)
            norm_ems.append(norm_em)

    raw_em_total = float(np.mean(raw_ems))
    norm_em_total = float(np.mean(norm_ems))

    return raw_em_total, norm_em_total

In [31]:
dev_raw_em, dev_norm_em = evaluate_bart(model_bart, tokenizer_bart, df_dev, BATCH_SIZE, MAX_SOURCE_LEN, MAX_TARGET_LEN)

print("=== Dev metrics ===\n")
print(f"Raw EM       : {dev_raw_em}")
print(f"Normalized EM: {dev_norm_em}")

=== Dev metrics ===

Raw EM       : 0.196
Normalized EM: 0.198


In [32]:
test_raw_em, test_norm_em = evaluate_bart(model_bart, tokenizer_bart, df_test, BATCH_SIZE, MAX_SOURCE_LEN, MAX_TARGET_LEN)

print("=== Test metrics ===\n")
print(f"Raw EM       : {test_raw_em}")
print(f"Normalized EM: {test_norm_em}")

=== Test metrics ===

Raw EM       : 0.193
Normalized EM: 0.196


In [33]:
df_dev_sample = df_dev.sample(5, random_state=1).reset_index(drop=True)

print("=== Qualitative Analysis on 5 dev examples ===\n")
for i, row in df_dev_sample.iterrows():
    question = row["question"]
    schema = row["schema"]
    label = row["query"]

    pred = pred_sql(model_bart, tokenizer_bart, [question], [schema], MAX_SOURCE_LEN, MAX_TARGET_LEN)

    raw_em, norm_em = compute_em_metrics(pred[0], label)

    print(f"--- Example {i + 1}--- ")
    print("Question:")
    print(question)
    print("\nModel prediction SQL:")
    print(pred[0])
    print("\nGold SQL:")
    print(label)
    print(f"\nRaw EM     : {raw_em}")
    print(f"Normalized EM: {norm_em}")
    print("-" * 80 + "\n")

=== Qualitative Analysis on 5 dev examples ===

--- Example 1--- 
Question:
What is the average fine for traffic violations in the city of San Francisco?

Model prediction SQL:
SELECT AVG(amount) FROM traffic_violations WHERE city = 'San Francisco';

Gold SQL:
SELECT AVG(amount) FROM traffic_violations WHERE city = 'San Francisco';

Raw EM     : 1
Normalized EM: 1
--------------------------------------------------------------------------------

--- Example 2--- 
Question:
Find artists who performed in multiple cities in 'artist_performances' table?

Model prediction SQL:
SELECT artist_id FROM artist_performances;

Gold SQL:
SELECT artist_id, COUNT(DISTINCT city_id) as unique_cities FROM artist_performances GROUP BY artist_id HAVING COUNT(DISTINCT city_id) > 1;

Raw EM     : 0
Normalized EM: 0
--------------------------------------------------------------------------------

--- Example 3--- 
Question:
What are the artifact types and their quantities for excavation sites in Greece?

Mode

<div dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی:</b><br>
<br>
مدل BART که متعلق به facebook است را می‌توان به‌عنوان نسخه‌ای دو‌بخشی از ترنسفورمر در نظر گرفت که یک بخش مخصوص فهم ورودی دارد و بخش دیگر مخصوص تولید خروجی. این معماری به شکل Encoder–Decoder طراحی شده است و پیش‌آموزش آن بر اساس بازسازی متن‌های تخریب‌شده انجام شده، یعنی ابتدا ورودی را به‌صورت ناقص یا به‌هم‌ریخته دریافت می‌کند و سپس تلاش می‌کند نسخه‌ی اصلی و سالم جمله را بازسازی کند. همین فرایند باعث می‌شود BART در دو مهارت اساسی قوی شود: یکی فهم دقیق متن و دیگری توانایی تولید خروجی روان و ساخت‌یافته.
<br>
  سمت Encoder رفتار شبیه BERT دارد و ورودی را با self-attention دوطرفه پردازش می‌کند؛ بنابراین هر کلمه در متن کامل‌ترین شکل ممکن در کانتکست قرار می‌گیرد و دنباله‌ی ورودی را به یک سری بردار متنی تبدیل می‌کند. سپس، هر بردار ورودی، پس از چند لایه از شبکه‌ های feed-forward، به یک نمایش متنی تبدیل می‌ شود که اطلاعاتی از کل توالی ورودی را شامل می‌شود. لذا هر بخش از پرسش و schema می‌تواند به تمام بخش‌های دیگر attend کند و مدل روابط بین سوال و schema را یاد بگیرد.
<br>
 سمت Decoder رفتار شبیه GPT دارد و خروجی را مرحله‌به‌مرحله و به‌صورت auto-regressive تولید می‌کند، اما در هر مرحله می‌تواند از طریق cross-attention به کل بردارهای تولیدشده در Encoder دسترسی داشته باشد.
<br>
در وظیفه‌ای مثل Text-to-SQL، ورودی توالی‌ای از متن طبیعی و schema پایگاه داده است و خروجی یک توالی جدید از جنس SQL است. این مسئله دقیقا همان فرمت دنباله-به-دنباله‌ای را دارد که BART برایش ساخته شده است. Encoder می‌تواند سوال و schema را با هم بخواند، روابط بین کلمات و ساختار داده‌ها را کشف کند و برای هر توکن یک نمایش معنایی غنی بسازد. Decoder هنگام تولید SQL، در هر گام می‌تواند بر اساس توکن‌هایی که تا آن لحظه ساخته و همین‌طور بر اساس اطلاعات رمزگذاری‌شده‌ی Encoder، تصمیم بگیرد چه ستون یا چه جدول یا چه کلیدواژه‌ای را بنویسد. این تفکیک واضح بین مرحله‌ی فهمیدن و مرحله‌ی نوشتن باعث می‌شود معماری Encoder–Decoder به‌صورت طبیعی برای کارهای ساختاری چون Text-to-SQL مناسب باشد.
<br>
مزیت دیگر BART این است که schemaهای واقعی معمولا طولانی، شامل چندین دستور CREATE TABLE، تعداد زیادی ستون و گاهی مجموعه‌ای از INSERTهای نمونه هستند. اگر قرار بود مدل بدون Encoder مجزا این حجم از توکن‌ها را در کنار سوال مدیریت کند، cognitive load رشد می‌کرد و مدل باید ورودی طولانی را با خروجی در یک دنباله‌ی واحد ترکیب می‌کرد. اما در BART تمام ورودی در Encoder ثبت می‌شود و Decoder فقط هر زمان لازم باشد به آن رجوع می‌کند. این طراحی باعث می‌شود حتی اگر schema بلند باشد، همچنان تولید SQL پایدار بماند.
<br>
پیش‌آموزش BART به‌صورت denoising نیز به درد این مسئله می‌خورد. چون مدل به‌طور طبیعی یاد گرفته بخش‌های گمشده‌ی متن را تکمیل کند و خروجی را منسجم نگه دارد، در ساخت یک query که ساختاری سخت‌گیر و حساس دارد، معمولا بهتر عمل می‌کند. در Text-to-SQL، کوچک‌ترین اشتباه در ترتیب کلمات یا نشانه‌گذاری می‌تواند query را نامعتبر کند؛ بنابراین مدل باید هم در فهم دقیق ورودی و هم در تولید خروجی بسیار سخت‌گیر و دقیق باشد، چیزی که معماری و پیش‌آموزش BART به‌طور همزمان فراهمش می‌کنند.
<br>

<br>
در ارزیابی Text-to-SQL، معیار EM یا Exact Match عملا قلب ارزیابی محسوب می‌شود؛ چون باید بررسی کند آیا مدل دقیقا همان query را تولید کرده که پاسخ صحیح است یا نه. اما دو نسخه از EM داریم: Raw EM و Normalized EM. هر دو یک ایده‌ی مشترک دارند، یعنی مقایسه‌ی دو رشته؛ اما در عمل رفتارشان کاملا متفاوت است و برای SQL نتایجشان تفاوت بزرگی ایجاد می‌کند.
<br>
Raw EM ساده‌ترین شکل مقایسه است. این معیار تنها دو رشته را بدون هیچ پردازشی کنار هم می‌گذارد. اگر خروجی مدل از نظر کاراکتر به کاراکتر، کاملا برابر با query درست باشد، مقدار EM برابر 1 است؛ در غیر این‌صورت برابر صفر است. این مقایسه همان‌قدر که ساده است، سخت‌گیر هم هست. کوچک‌ترین تفاوت، حتی یک فاصله‌ی اضافه، یک newline متفاوت، تفاوت در استفاده از حروف کوچک و بزرگ، سمیکالن ته جمله، یا جابه‌جایی توکن‌هایی که از نظر معنایی بی‌اهمیت‌اند، باعث می‌شود Raw EM این دو رشته را کاملا متفاوت حساب کند. در زبان طبیعی این سخت‌گیری ممکن است تا حدی قابل‌قبول باشد، اما در SQL معمولا ناکافی است، چون SQL زبانی است که در آن ظاهر می‌تواند بدون تغییر معنای دستور تغییر کند. یعنی ممکن است دو query دقیقا همان خروجی را تولید کنند، اما به‌دلیل تفاوت‌های سطحی در شکل نوشتار، Raw EM آن‌ها را متفاوت بشمارد. بنابراین Raw EM اگرچه دقیق است، اما عادلانه نیست و برای Text-to-SQL معمولا شاخص سخت‌گیرانه‌ای محسوب می‌شود.
<br>
Normalized EM برای حل همین مشکل به‌وجود آمده. این معیار قبل از مقایسه‌ی دو query، آن‌ها را normalize می‌کند؛ یعنی جنبه‌هایی از نوشتار را که تأثیری در اجرای SQL ندارند حذف یا یکسان‌سازی می‌کند. معمولا فاصله‌های اضافی حذف می‌شوند، چند فاصله‌ی پشت‌سرهم به یک فاصله تبدیل می‌شود، حروف کوچک و بزرگ به یک حالت (غالبا lowercase) تبدیل می‌شوند، سمیکالن پایانی حذف می‌شود، و فضای خالی ابتدا و انتهای query پاک می‌شود. بعد از این پردازش، دو query که شاید در ظاهر متفاوت بودند، ولی از نظر معنا و syntax معتبر و یکسان هستند، کنار هم یکسان به نظر می‌رسند و خروجی EM برای آن‌ها برابر 1 خواهد بود. نقش Normalized EM این است که مقایسه را از سطح کاراکترهای ظاهری به سطح ساختار واقعی query نزدیک کند و میزان حساسیت را تنها روی چیزهایی متمرکز کند که واقعا اهمیت دارند؛ مثلا ترتیب ستون‌ها یا بخش‌ها، وجود شرط‌های درست، یا انتخاب جدول صحیح.
<br>
به‌صورت خلاصه، Raw EM یک معیار باینری کاملا ماشینی است که فقط شکل ظاهری را می‌بیند و برای SQL اغلب بیش از حد سخت‌گیرانه است، درحالی‌که Normalized EM با حذف عوامل بی‌ربط و متمرکز کردن مقایسه روی ساختار و معنا، تصویری منصفانه‌تر و دقیق‌تر از توانایی مدل در حل مسئله‌ی Text-to-SQL ارائه می‌دهد. این دو معیار در کنار هم کمک می‌کنند بفهمیم مدل فقط قالب نوشتاری SQL را رعایت نکرده، یا اساسا ساختار درست query را یاد نگرفته است؛ تفاوتی که برای تحلیل رفتار مدل بسیار حیاتی است.
<br>

<br>
طبق پنج نمونه‌ی مشاهده شده در پایان، رفتار مدل BART خیلی خوب نشان می‌دهد که کجا مسئله را واقعا می‌فهمد و کجا عملا فقط شبیه‌سازی SQL انجام می‌دهد بدون این‌که منطق دقیق سوال و schema را تا آخر دنبال کند. در ساده‌ترین حالت‌ها، جایی که سوال یک کوئری تجمیعی تک‌جدولی با یک شرط ساده است، مدل کاملا موفق عمل کرده است. در مثال‌های اول و چهارم، هم پرسش واضح است، هم schema ساده است و هم نگاشت زبانی به SQL نسبتا مستقیم است؛ نتیجه این شده که کوئری پیش‌بینی‌شده توسط مدل و کوئری طلایی از نظر Raw EM و Normalized EM هر دو برابر 1 هستند. این یعنی مدل یاد گرفته الگوی کلی SELECT agg(column) FROM table WHERE condition را خوب تقلید کند و در چنین سناریوهای خطی و کم‌ابهام دقتش بالاست.
<br>
در مقابل، همین‌که ساختار منطقی سوال کمی پیچیده‌تر می‌شود، ضعف‌های جدی ظاهر می‌شوند. در مثال دوم، صورت سوال به‌وضوح به شمارش تعداد شهرهای متمایز برای هر هنرمند اشاره می‌کند، یعنی نیاز به GROUP BY artist_id و COUNT(DISTINCT city_id) و محدود کردن خروجی به آن‌هایی که بیش از یک شهر دارند. کوئری طلایی دقیقا همین را پیاده کرده است. اما مدل به‌جای رفتن سراغ مفهوم multiple cities از طریق aggregation و having، عملا کل ساختار مسئله را ساده کرده و تنها یک SELECT ساده بدون هیچ GROUP BY و بدون هیچ شرط مرتبطی تولید کرده است. این نشان می‌دهد مدل نه‌تنها الگوی group by + count + having را درونی نکرده، بلکه در مواجهه با مفاهیم کمی پیچیده‌تر، تمایل دارد مسئله را به ساده‌ترین حالت قابل تولید فروبکاهد. این نوع خطا یک semantic error کامل است؛ یعنی syntax به‌ظاهر درست است اما پیوندی با نیت سوال ندارد.
<br>
در مثال سوم، مسئله یک پله‌ی دیگر هم پیچیده‌تر می‌شود؛ این‌جا هم join لازم است، هم aggregation و هم group by روی نوع آرتیفکت. کوئری طلایی دقیقا از جدول Artifact با alias a و ExcavationSite با alias e استفاده کرده، روی SiteID join کرده، کشور یونان را فیلتر کرده و بر حسب a.Type گروه‌بندی کرده است. مدل اما کوئری‌ای تولید کرده که چندین ایراد ساختاری و معنایی هم‌زمان دارد: هم از aliasها به‌صورت نامنظم و ناسازگار استفاده کرده، هم join را روی ستون‌های اشتباه بسته، هم group by را به شکلی نادرست به کار برده، و هم حتی فیلدی مانند TotalQuantity را اضافه کرده که ذاتا در schema وجود ندارد. خروجی مدل از نظر ساختار SQL معتبر به نظر می‌رسد اما از نظر معنی، نه با schema سازگار است و نه با منطق مسئله. چنین خطاهایی نشان می‌دهد که مدل در مدیریت join، alias و تطبیق دقیق schema با پرسش همچنان شکننده است و در بسیاری مواقع به‌جای بازسازی دقیق یک الگوی صحیح، ترکیبی از الگوهای دیده‌شده در پیش‌آموزش را به‌صورت کنارهم‌چینی تصادفی تولید می‌کند.
<br>
در مثال پنجم، مدل بخشی از اطلاعات خواسته‌شده را درک کرده اما شرط اصلی مسئله را نادیده گرفته است. پرسش صراحتا درباره‌ی فروش خودروهای برقی در آلمان و بریتانیا از سال ۲۰۲۰ به بعد است، ولی مدل تنها قسمت مربوط به کشورهای Germany و UK را لحاظ کرده و شرط year >= 2020 را به‌طور کامل حذف کرده است. در ظاهر، اختلاف نحوی بین IN و (country = ... OR country = ...) کم‌اهمیت است، اما حذف شرط سال باعث می‌شود خروجی مدل به‌طور معنی‌داری با نیت پرسش فاصله داشته باشد، در حالی‌که کوئری طلایی دقیقا همین قید را هدف گرفته است. این نمونه نشان می‌دهد که در پرسش‌های چندشرطی، مدل به‌ویژه در انتخاب و نگهداری تمام قیود مسئله دچار لغزش می‌شود.
<br>
در مجموع، این پنج مثال تصویری نسبتا شفاف از رفتار مدل برای Text-to-SQL می‌سازند. در کوئری‌های تک‌جدولی ساده که فقط یک aggregation و یک شرط مستقیم دارند، مدل کاملا خوب عمل می‌کند و هم syntax و هم semantics را درست می‌سازد. اما در پرسش‌هایی که نیاز به reasoning سطح بالاتر دارند، مثل تشخیص مفهوم چند شهر، مدیریت join بین جداول، استفاده صحیح از alias و پیاده‌سازی group by، مدل دچار اختلال می‌شود و تمایل دارد ساختار مسئله را بیش از حد ساده کند یا بخش‌هایی را بی‌منطق اضافه و حذف کند. این رفتار نشان می‌دهد که BART در این سطح آموزش، هنوز ترجمه‌ی ساختاریافته‌ی سوال به SQL را به شکل کامل یاد نگرفته و در بسیاری موارد بر الگوهای سطحی تکیه می‌کند. با این حال، نتایج نشان می‌دهند که مدل در بخش‌هایی از مسئله توانایی دارد و با آموزش بیشتر یا constraintهای سخت‌گیرانه‌تر در زمان decode می‌تواند عملکرد پایدارتری ارائه دهد.
</div>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش سوم: آموزش مدل Decoder-only (GPT-2) (۱۵ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش تسک Text-to-SQL را با معماری GPT-2 پیاده‌سازی می‌کنید. ابتدا قالب ورودی مناسب برای مدل‌های decoder-only طراحی کنید، به‌طوری‌که پرسش و schema به‌صورت یک prefix مشخص (مانند question: … schema: … SQL:) به مدل داده شوند و مدل ادامه آن را به‌صورت SQL تولید کند. سپس یک Dataset مخصوص causal LM بسازید که full sequence (prefix + gold SQL + EOS) را تولید کرده و بخش prefix را در برچسب‌ها mask کند. مدل GPT-2 را یک epoch آموزش دهید و loss و زمان اجرا را گزارش کنید. هنگام ارزیابی، prefix را از خروجی حذف کنید و تنها SQL تولیدشده را بسنجید. معیارهای Raw EM و Normalized EM را روی dev و test گزارش کرده و ۵ نمونه از خروجی مدل را بررسی و تحلیل کنید.
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک مدل GPT-2 آموزش‌دیده و ارزیابی‌شده است که با استفاده از قالب ورودی مبتنی بر prefix قادر به تولید SQL بوده و معیارهای Raw EM و Normalized EM آن روی مجموعه‌های dev و test به‌درستی محاسبه و گزارش شده باشند.
</p>

In [34]:
from transformers import GPT2Tokenizer, GPT2LMHeadModel

In [35]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

print(device)

cuda


In [57]:
class TextToSQLGPT2Dataset(Dataset):
    def __init__(self, df, tokenizer, max_seq_len=512):
        self.df = df.reset_index(drop=True)

        self.tokenizer = tokenizer

        self.max_seq_len = max_seq_len

    def __len__(self):
        return len(self.df)


    def _build_prompt(self, question, schema):
        return f"Question: {question}\nSchema: {schema}\nSQL:"

    def __getitem__(self, idx):
        row = self.df.iloc[idx]

        question = row["question"]
        schema = row["schema"]
        query = row["query"]

        prefix = self._build_prompt(question, schema)

        full_text = prefix + " " + query + self.tokenizer.eos_token

        seq_encode = self.tokenizer(
            full_text,
            max_length=self.max_seq_len,
            truncation=True,
            padding="max_length",
            return_tensors="pt",
        )

        prefix_ids = self.tokenizer(
            prefix,
            add_special_tokens=False
        ).input_ids

        input_ids = seq_encode["input_ids"].squeeze(0)
        attention_mask = seq_encode["attention_mask"].squeeze(0)
        labels = input_ids.clone()

        len_prefix = min(len(prefix_ids), self.max_seq_len)
        labels[ : len_prefix] = -100

        return {
            "input_ids": input_ids,
            "attention_mask": attention_mask,
            "labels": labels,
        }

In [58]:
MODEL_NAME_GPT2 = "gpt2"

tokenizer_gpt2 = GPT2Tokenizer.from_pretrained(MODEL_NAME_GPT2)
tokenizer_gpt2.pad_token = tokenizer_gpt2.eos_token

model_gpt2 = GPT2LMHeadModel.from_pretrained(MODEL_NAME_GPT2)
model_gpt2.resize_token_embeddings(len(tokenizer_gpt2))
model_gpt2.to(device)

GPT2LMHeadModel(
  (transformer): GPT2Model(
    (wte): Embedding(50257, 768)
    (wpe): Embedding(1024, 768)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-11): 12 x GPT2Block(
        (ln_1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (attn): GPT2Attention(
          (c_attn): Conv1D(nf=2304, nx=768)
          (c_proj): Conv1D(nf=768, nx=768)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): GPT2MLP(
          (c_fc): Conv1D(nf=3072, nx=768)
          (c_proj): Conv1D(nf=768, nx=3072)
          (act): NewGELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
  (lm_head): Linear(in_features=768, out_features=50257, bias=False)
)

In [59]:
MAX_SEQ_LEN = 512

dataset_train = TextToSQLGPT2Dataset(df_train, tokenizer_gpt2, MAX_SEQ_LEN)
dataset_dev = TextToSQLGPT2Dataset(df_dev, tokenizer_gpt2, MAX_SEQ_LEN)
dataset_test = TextToSQLGPT2Dataset(df_test, tokenizer_gpt2, MAX_SEQ_LEN)

In [64]:
BATCH_SIZE = 4
TRAIN_EPOCHS = 1
LEARNING_RATE = 5e-5

training_args = TrainingArguments(
    output_dir="./gpt2_text2sql",
    num_train_epochs=TRAIN_EPOCHS,
    per_device_train_batch_size=BATCH_SIZE,
    per_device_eval_batch_size=BATCH_SIZE,
    learning_rate=LEARNING_RATE,
    weight_decay=0.01,
    warmup_ratio=0.05,
    fp16=False,
    logging_strategy="steps",
    logging_steps=100,
    report_to="none"
)

In [68]:
trainer = Trainer(
    model=model_gpt2,
    args=training_args,
    train_dataset=dataset_train,
    eval_dataset=dataset_dev,
    data_collator=default_data_collator,
    tokenizer=tokenizer_gpt2
)

/tmp/ipykernel_47/292046365.py:1: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [69]:
start_time = time.time()
train_result = trainer.train()
end_time = time.time()

final_train_loss = train_result.training_loss
train_metrics = train_result.metrics

print("\n=== Training Report ===")
print(f"\nFinal training loss: {final_train_loss}")
print("\nTraining time (seconds):", end_time - start_time)
print("\nTrain metrics:", train_metrics)
print("\nHyperparameters:")
print(f"  num_train_epochs: {training_args.num_train_epochs}")
print(f"  per_device_train_batch_size: {training_args.per_device_train_batch_size}")
print(f"  max_seq_len: {MAX_SEQ_LEN}")
print(f"  learning_rate: {training_args.learning_rate}")
print(f"  weight_decay: {training_args.weight_decay}")
print(f"  warmup_ratio: {training_args.warmup_ratio}")


=== Training Report ===

Final training loss: 0.11697582893371582

Training time (seconds): 2515.3521473407745

Train metrics: {'train_runtime': 2514.8833, 'train_samples_per_second': 7.953, 'train_steps_per_second': 1.988, 'total_flos': 5225840640000000.0, 'train_loss': 0.11697582893371582, 'epoch': 1.0}

Hyperparameters:
  num_train_epochs: 1
  per_device_train_batch_size: 4
  max_seq_len: 512
  learning_rate: 5e-05
  weight_decay: 0.01
  warmup_ratio: 0.05


In [83]:
def pred_sql(model, tokenizer, question, schema, max_new_toks=512):
    prefix = f"Question: {question}\nSchema: {schema}\nSQL:"

    prefix_encode = tokenizer(
        prefix,
        add_special_tokens=False,
        return_tensors="pt"
    ).to(model.device)

    input_ids = prefix_encode["input_ids"]
    attention_mask = prefix_encode["attention_mask"]

    with torch.no_grad():
        generated_ids = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask,
            max_new_tokens=max_new_toks,
            num_beams=4,
            pad_token_id=tokenizer.eos_token_id
        )

    continuation_ids = generated_ids[0][input_ids.shape[1] : ]

    pred_sqls = tokenizer.decode(
        continuation_ids,
        skip_special_tokens=True
    )

    return pred_sqls

In [85]:
def evaluate_gpt2(model, tokenizer, df, max_new_toks=512):
    raw_ems = []
    norm_ems = []

    for i, row in df.iterrows():
        question = row["question"]
        schema = row["schema"]
        label = row["query"]

        pred = pred_sql(model, tokenizer, question, schema, max_new_toks)

        raw_em, norm_em = compute_em_metrics(pred, label)

        raw_ems.append(raw_em)
        norm_ems.append(norm_em)

    raw_em_total = float(np.mean(raw_ems))
    norm_em_total = float(np.mean(norm_ems))

    return raw_em_total, norm_em_total

In [81]:
MAX_NEW_TOKS = 128

In [80]:
dev_raw_em, dev_norm_em = evaluate_gpt2(model_gpt2, tokenizer_gpt2, df_dev, MAX_NEW_TOKS)

print("=== Dev metrics ===\n")
print(f"Raw EM       : {dev_raw_em}")
print(f"Normalized EM: {dev_norm_em}")

=== Dev metrics ===

Raw EM       : 0.164
Normalized EM: 0.164


In [86]:
test_raw_em, test_norm_em = evaluate_gpt2(model_gpt2, tokenizer_gpt2, df_test, MAX_NEW_TOKS)

print("=== Test metrics ===\n")
print(f"Raw EM       : {test_raw_em}")
print(f"Normalized EM: {test_norm_em}")

=== Test metrics ===

Raw EM       : 0.1675
Normalized EM: 0.171


In [87]:
df_dev_sample = df_dev.sample(5, random_state=1).reset_index(drop=True)

print("=== Qualitative Analysis on 5 dev examples ===\n")
for i, row in df_dev_sample.iterrows():
    question = row["question"]
    schema = row["schema"]
    label = row["query"]

    pred = pred_sql(model_gpt2, tokenizer_gpt2, question, schema, MAX_NEW_TOKS)

    raw_em, norm_em = compute_em_metrics(pred, label)

    print(f"--- Example {i + 1}--- ")
    print("Question:")
    print(question)
    print("\nModel prediction SQL:")
    print(pred)
    print("\nGold SQL:")
    print(label)
    print(f"\nRaw EM     : {raw_em}")
    print(f"Normalized EM: {norm_em}")
    print("-" * 80 + "\n")

=== Qualitative Analysis on 5 dev examples ===

--- Example 1--- 
Question:
What is the average fine for traffic violations in the city of San Francisco?

Model prediction SQL:
 SELECT AVG(amount) FROM traffic_violations WHERE city = 'San Francisco';

Gold SQL:
SELECT AVG(amount) FROM traffic_violations WHERE city = 'San Francisco';

Raw EM     : 1
Normalized EM: 1
--------------------------------------------------------------------------------

--- Example 2--- 
Question:
Find artists who performed in multiple cities in 'artist_performances' table?

Model prediction SQL:
 SELECT DISTINCT artist_id, city_id FROM artist_performances WHERE date BETWEEN DATE_SUB(CURRENT_DATE, INTERVAL 1 YEAR) AND CURRENT_DATE;

Gold SQL:
SELECT artist_id, COUNT(DISTINCT city_id) as unique_cities FROM artist_performances GROUP BY artist_id HAVING COUNT(DISTINCT city_id) > 1;

Raw EM     : 0
Normalized EM: 0
--------------------------------------------------------------------------------

--- Example 3--- 


<div dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی:</b><br>
<br>
طبق پنج نمونه‌ی مشاهده‌شده در پایان، رفتار GPT-2 در مقایسه با BART نشان می‌دهد که این مدل در مواجهه با ساختارهای ساده‌ی Text-to-SQL عملکرد قابل‌قبولی دارد، اما به محض این‌که منطق سوال پیچیده‌تر می‌شود، تولید آن حالت زبانی و الگویی پیدا می‌کند و از ساختار واقعی schema فاصله می‌گیرد. در مثال اول، پرسش کاملا مستقیم و از جنس یک کوئری تجمیعی ساده است و GPT-2 بدون مشکل الگوی SELECT AVG(...) FROM ... WHERE ... را تکرار کرده و دقیقا همان چیزی را ساخته که در داده‌ی طلایی آمده؛ بنابراین Raw و Normalized EM هر دو برابر ۱ شده‌اند. در مثال چهارم هم همین الگوی موفق دیده می‌شود؛ سوال فقط یک SUM روی یک جدول با یک شرط ساده‌ی equality است و مدل نیز همین ساختار را درست تولید کرده است. این موارد نشان می‌دهند که GPT-2 در بازتولید الگوهای پرتکرار SQL، به‌ویژه وقتی با یک prefix واضح تغذیه می‌شود، مشکلی ندارد و می‌تواند پرسش‌هایی با پیچیدگی کم را به‌خوبی به SQL متناظر ترجمه کند.
<br>
اما همین‌که سوال ساختار منطقی پیچیده‌تری پیدا می‌کند، ضعف‌های بنیادی GPT-2 نمایان می‌شوند. مثال دوم درباره‌ی هنرمندانی است که در چند شهر اجرا داشته‌اند؛ مسئله‌ای که ذاتا به یک group by روی artist_id، یک COUNT(DISTINCT city_id) و یک شرط HAVING برای محدود کردن پاسخ‌ها نیاز دارد. کوئری طلایی همین الگو را پیاده کرده است. در مقابل، GPT-2 به‌جای این ساختار، یک کوئری از جنس DISTINCT همراه با یک فیلتر زمانی ساخته است؛ یعنی نه aggregation انجام داده، نه group by، و نه کوچک‌ترین ارتباطی با مفهوم اجرا در چند شهر برقرار کرده است. این نوع انحراف نشان می‌دهد که GPT-2 درک استدلالی درباره‌ی ارتباط پرسش و schema ندارد و بیشتر بر اساس الگوهای متنی که در پیش‌آموزش دیده چنین خروجی‌هایی را می‌سازد.
<br>
در مثال سوم، این فاصله میان نیاز مسئله و خروجی مدل شدیدتر می‌شود. سوال به‌طور مشخص درباره‌ی انواع آرتیفکت‌ها و تعدادشان در سایت‌های حفاری یونان است. کوئری طلایی با یک join تمیز بین Artifact و ExcavationSite و یک group by بر اساس نوع آرتیفکت این هدف را محقق می‌کند. اما GPT-2 به‌جای این ساختار روشن، چندین join تکراری و بی‌ارتباط تولید کرده، aliasها را به‌صورت ناسازگار و چندباره استفاده کرده و ستون‌هایی مانند ArtifactName و ArtifactQuantity را وارد کوئری کرده که در صورت سوال و کوئری طلایی وجود ندارند. خروجی حاصل عملا ترکیبی از الگوهای SQL آشناست که بدون درک واقعی از schema به هم چسبیده‌اند؛ نمونه‌ای مشخص از اشتباه ساختاری که در مدل‌های صرفا دیکودر زیاد دیده می‌شود.
<br>
در مثال پنجم، GPT-2 توانسته بخشی از مسئله را به‌درستی درک کند؛ شرط کشورها را با IN پیاده کرده و ساختار WHERE کلی قابل‌قبول است. اما شرط اصلی سوال یعنی year >= 2020 را یا درک نکرده یا نتوانسته درست ترجمه کند و در عوض year IN (2020) را ساخته است؛ شرطی که فقط یک سال را پوشش می‌دهد و هیچ‌یک از سال‌های بعد را شامل نمی‌شود. این نمونه نشان می‌دهد که مدل حتی وقتی به ساختار کلی نزدیک می‌شود، در انتخاب اپراتورهای درست برای بیان بازه‌ها ضعف دارد و ترجیح می‌دهد فرم آشناتر IN را جایگزین یک شرط بازه‌ای کند.
<br>
در مجموع، این پنج مثال نشان می‌دهند که GPT-2 در کوئری‌های ساده و تک‌جدولی عملکرد معقولی دارد و می‌تواند الگوهای مستقیم SQL را از روی prefix ادامه دهد. اما در پرسش‌هایی که نیاز به reasoning ساختاری، ارتباط دقیق با schema، پیاده‌سازی درست join یا استفاده‌ی صحیح از aggregation و group by دارند، مدل دچار خطاهای اساسی می‌شود. خروجی‌ها در این شرایط یا ساده‌سازی بیش از حد مسئله هستند یا ترکیبی از الگوهای زبانی آشنا که هیچ هم‌خوانی معنایی با هدف پرسش ندارند. این رفتار کاملا مطابق انتظار برای یک معماری decoder-only است که به‌جای تحلیل ورودی، تنها ادامه‌ی محتمل‌ترین رشته‌ی ممکن را تولید می‌کند و بنابراین در وظایف translationگونه‌ی ساختاریافته مثل Text-to-SQL ذاتا محدودتر از معماری‌های encoder–decoder ظاهر می‌شود.
</div>


### <div style="text-align: center; direction: rtl; font-family: Vazir;">بخش چهارم: مقایسه و تحلیل نهایی (۱۰ نمره)</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این بخش باید عملکرد BART و GPT-2 را از جنبه‌های مختلف مقایسه کنید. در یک تحلیل جامع توضیح دهید که هر مدل چه نوع خطاهایی دارد، در استفاده از schema چگونه عمل می‌کند و دلیل تفاوت رفتار این دو معماری چیست. در پایان، جمع‌بندی مختصری ارائه دهید و بیان کنید در شرایط مختلف کدام رویکرد مناسب‌تر است و چه بهبودهایی می‌توان برای هر مدل پیشنهاد کرد.</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
خروجی مورد انتظار این بخش، یک تحلیل جامع و مقایسه‌محور است که در آن عملکرد، خطاها، نقاط قوت و ضعف دو مدل BART و GPT-2 بر اساس معیارهای ارزیابی، نمونه‌های تولیدی و تفاوت‌های معماری به‌طور دقیق بررسی و جمع‌بندی شده باشد.
</p>

<div dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی:</b><br>
<br>
طبق نتایج به‌دست‌آمده از هر دو مدل، هم در معیارهای عددی و هم در نمونه‌های کیفی، تفاوت رفتار BART و GPT-2 در Text-to-SQL کاملا قابل مشاهده است. در سطح کلی، اعداد EM نشان می‌دهند که هر دو مدل همچنان از سطح ایده‌آل فاصله دارند، اما معماری Encoder–Decoder در BART اندکی پایدارتر و دقیق‌تر عمل کرده است. امتیازهای BART روی dev و test حدود 0.19 برای Raw EM و حدود 0.196–0.198 برای Normalized EM هستند، در حالی که GPT-2 در dev حدود 0.164 و در test حدود 0.1675 قرار دارد. اختلاف زیاد نیست، اما ثبات BART در هر دو مجموعه‌داده کمی بهتر است و همین اختلاف کوچک هم با نمونه‌های کیفی کاملا هم‌خوانی دارد.
<br>
رفتار BART نشان می‌دهد که این مدل در مواجهه با کوئری‌های ساده‌ی تک‌جدولی، به‌خوبی می‌تواند نگاشت مستقیم سوال به SQL را اجرا کند. در نمونه‌ها، سوالاتی مانند میانگین جریمه یا مجموع ذخایر copper بدون خطا تولید شده‌اند؛ دقیقا همان‌طور که معیارهای EM کامل نشان می‌دهند. با این حال، وقتی ساختار سوال پیچیده‌تر می‌شود، مثل نیاز به group by، aggregationهای تو در تو، انتخاب صحیح alias یا مدیریت join بین دو جدول، BART دچار لغزش می‌شود. خطاهای آن معمولا از جنس ساده‌سازی بیش از حد مسئله‌اند؛ مثلا در مثال چند شهر اجرا، کل سازوکار group by و having را کنار می‌گذارد و یک SELECT ساده تولید می‌کند. یا در مثال آرتیفکت‌ها، اگرچه ساختار join را تشخیص می‌دهد و از هر دو جدول استفاده می‌کند، اما عملا منطق سوال را تا انتها نمی‌بندد و یا aggregation و group by را اشتباه پیاده می‌کند. این نوع خطاها نشان می‌دهند که BART schema را می‌بیند و از اطلاعات آن استفاده می‌کند، اما رابطه‌ی دقیق بین سوال و ساختار پایگاه‌داده را همیشه به‌طور کامل بازسازی نمی‌کند.
<br>
در مقابل، GPT-2 به‌عنوان یک معماری کاملا decoder-only، رفتار متفاوتی دارد. مثل BART، در کوئری‌های ساده عملکرد عالی دارد و نمونه‌های موفق آن کاملا بی‌نقص هستند. اما در کوئری‌های پیچیده‌تر، نوع خطاهایش از جنس دیگری است: GPT-2 معمولا به سمت الگوهای زبانی آشنا و پرتکرار از پیش‌آموزش می‌رود، بدون آن‌که ضرورتا از schema یا منطق پرسش استفاده کند. در مثال هنرمندان با اجرا در چند شهر، به جای group by و count(distinct)، یک فیلتر زمانی بی‌ربط روی date تولید کرد؛ در نمونه‌ی مربوط به آرتیفکت‌ها هم ترکیبی از چندین join تکراری و self-joinهای اشتباه ساخت که عملا نشان می‌دهد مدل اسکیمای جدول را درک نکرده و صرفا الگوهای زبانی SQL را به‌صورت ترکیبی بازتولید کرده است. خطای GPT-2 بیشتر شبیه ترکیب نامنظم پترن‌های SQL است که از نظر نحوی مشابه SQL به‌نظر می‌رسد اما از نظر منطقی ناقص یا کاملا بی‌ربط است.
<br>
تفاوت اصلی این دو مدل در معماری آن‌ها ریشه دارد. در BART، encoder یک نمایش غنی از سوال و schema می‌سازد و decoder می‌تواند در هر مرحله‌ی تولید، روی بخش‌های مختلف ورودی تمرکز کند. همین توجه دوسویه به سوال و ساختار پایگاه‌داده باعث می‌شود BART درک نسبی از رابطه‌ی بین آن‌ها داشته باشد، حتی اگر خروجی نهایی همیشه درست نباشد. در GPT-2، ورودی و خروجی یک دنباله‌ی واحد هستند و مدل هیچ بخش مستقل و تخصصی برای درک سوال ندارد؛ تنها کاری که انجام می‌دهد ادامه دادن متن است. بنابراین وقتی پرسش پیچیده می‌شود، GPT-2 به‌جای استدلال، الگوهای زبانی آشنا را به‌صورت تداعی‌وار تولید می‌کند و همین باعث می‌شود در مسائل چندجدولی یا شامل aggregationهای خاص، عملکردش نامطمئن‌تر از BART باشد.
<br>
در مورد نقاط قوت دو مدل، هر دو از این جهت قابل توجه‌اند که در کوئری‌های ساده و تک‌جدولی با ساختار خطی عملکرد بسیار پایداری نشان می‌دهند و می‌توانند الگوهای پرتکرار SQL را به‌خوبی بازتولید کنند. BART به‌دلیل داشتن encoder می‌تواند اطلاعات سوال و schema را به‌صورت منسجم درک کند و در بسیاری از موارد، حتی در کوئری‌هایی که یک سطح بالاتر از سادگی هستند، ساختار کلی جدول‌ها، نام ستون‌ها و سرعت نگاشت پرسش به SQL را بهتر رعایت کند. GPT-2 نیز در متونی که شباهت زیادی به داده‌های پیش‌آموزشش دارند، خروجی روان، نحوی و از نظر زبانی سازگار تولید می‌کند و در بازتولید الگوهای syntax آشنا عملکرد چشم‌گیری دارد. هر دو مدل همچنین قادرند در شرایطی که پرسش و schema به‌صورت واضح و کوتاه بیان شده باشند، با حداقل آموزش نیز به کاربری عملی نزدیک شوند.
<br>
در مقابل، نقاط ضعف هر دو مدل عمدتا زمانی آشکار می‌شوند که مسئله نیازمند استدلال ساختاری، استفاده‌ی پیچیده از schema، یا پیاده‌سازی کوئری‌های چندمرحله‌ای باشد. BART با وجود درک بهتر ورودی، هنوز در هماهنگ کردن تمام اجزای کوئری، aliasها، group by، join و شروط چندگانه، دچار خطای منطقی می‌شود و گاهی مسئله را بیش از حد ساده می‌کند. GPT-2 نیز به‌دلیل نبود encoder، در مواجهه با پیچیدگی‌ها به الگوهای زبانی آشنا پناه می‌برد و همین باعث تولید joinهای بی‌ربط، شروط اضافی یا ساختارهای تکراری و نادرست می‌شود. هر دو مدل به‌طور کلی نسبت به تنوع بیان منطقی در SQL حساس‌اند و عدم توازن میان سوال و schema می‌تواند خروجی را نامعتبر کند. بنابراین اگرچه توانایی‌های اولیه‌ی این دو مدل قابل استفاده است، اما برای مسائل واقعی و پیچیده‌ی Text-to-SQL که نیازمند دقت معنایی بالا و استدلال مبتنی بر schema هستند، هر دو مدل نیازمند تقویت، محدودکننده‌ی نحوی و راهبردهای پیشرفته‌ی decode می‌باشند.
<br>
در جمع‌بندی، برای وظایف Text-to-SQL که به استفاده‌ی دقیق از schema، مدیریت join، group by و نگاشت ساختاریافته‌ی اطلاعات نیاز دارند، معماری Encoder–Decoder مانند BART مناسب‌تر است؛ زیرا از نظر معماری توانایی پردازش جداگانه‌ی ورودی و سپس تولید منسجم خروجی را دارد. در مقابل، GPT-2 در سوالات ساده و خطی، به‌ویژه زمانی که معادل‌شان مستقیما قابل حدس زدن از روی الگوهای زبانی باشد، می‌تواند به‌خوبی عمل کند، اما با افزایش پیچیدگی منطق سوال، احتمال خطا و تولید ساختارهای نامعتبر در آن افزایش پیدا می‌کند. برای بهبود BART، می‌توان از آموزش طولانی‌تر، محدودکننده‌های نحوی در decoding، مدل‌های دستیار برای بررسی سازگاری schema، و روش‌هایی مثل PICARD استفاده کرد. برای GPT-2 نیز می‌توان prompting دقیق‌تر، محدودیت‌های سخت‌گیرانه روی دستور زبان SQL، یا ترکیب آن با یک ماژول کنترلی جهت جلوگیری از تولید الگوهای ساختاری غلط را پیشنهاد داد. در نهایت، اگر هدف کاربردی و مقیاس بزرگ باشد، معماری Encoder–Decoder به‌طور تجربی و مفهومی گزینه‌ی مؤثرتر و قابل‌اعتماد‌تری است.
</div>


# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>


<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA4_Q1_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
